### PR TITLE
MGTwoLevelTransfer: improve setup of indices

### PR DESCRIPTION
**Pull request by peterrum**
references https://github.com/dealii/dealii/issues/14968

### CODE REVIEW

File: include/deal.II/matrix_free/constraint_info.h [hunk 1]
```diff
@@ -85,15 +85,29 @@ namespace internal
...
+      template <typename T = unsigned int>
       void
       reinit(const DoFHandler<dim> &dof_handler,
              const unsigned int     n_cells,
              const bool             use_fast_hanging_node_algorithm = true);
 
+      template <typename T = unsigned int>
       void
       read_dof_indices(
         const unsigned int                                    cell_no,
```
Comment (bangerth) [inline include/deal.II/matrix_free/constraint_info.h:113]: I have nothing of consequence to say about the patch itself other than
* You'll want to document this function.
* You'll want to document what `T` stands for (for all of these functions). There may also be a better name than just `T`.
* Giving `T` a default value invites people to omit the argument altogether, and perhaps introduce a bug in the process.

File: include/deal.II/matrix_free/constraint_info.h [hunk 2]
```diff
@@ -176,6 +204,35 @@ namespace internal
...
+      template <typename T>
+      std::vector<std::vector<T>> &
+      get_dof_indices_per_cell()
+      {
+        if constexpr (std::is_same_v<T, unsigned int>)
+          return dof_indices_per_cell;
+        else if constexpr (std::is_same_v<T, types::global_dof_index>)
+          return dof_indices_per_cell_long;
+        else
+          return {};
```
Comment (bangerth) [inline include/deal.II/matrix_free/constraint_info.h:222]: Is that last `else` what you want? Or do you really want to ensure that `T` can only either be `unsigned int` or `types::global_dof_index`?

**Comment by kronbichler** [comment]
Can you explain what you're trying to achieve here and what the underlying algorithm is? I see you're creating duplicated data structures (with some TODOs left in the code) to keep both `unsigned int` and `types::global_dof_index` variables in the constraint info class, in order to be able to keep some indices in global index space before a partitioner can be set up (because we cannot easily come up with a number fitting in `unsigned int` for arbitrary ghost indices). Is that right? I am wondering if there any other way to achieve the same, e.g. templating the whole class for the temporary setup where we need `types::global_dof_index` for the intermediate ghost data? Even though I have a fairly good understanding of the class `mg_transfer_global_coarsening.templates.h`, I must admit that the changes are confusing and poorly documented. As an aside, having the ability to return different data structures from the same surrounding object is inviting for mistakes.

**Comment by kronbichler** [comment]
On a second thought, I believe this PR seems to replicate what is done in https://github.com/dealii/dealii/blob/048ef4f4fab0e0a194ed6c28402babaf88411d50/include/deal.II/matrix_free/dof_info.templates.h#L150-L159 (and several more places in that file), but in a somewhat different way. I think this illustrates that we should aim to unify `DoFInfo` and `ConstraintInfo` at some point, issue #15162, because having duplicate data structures doing the same thing has a cost, and having different implementations for doing that same thing (because the people writing it have different tastes, or because we did bother to take the time to make a good interface) has an even higher cost. This does not mean that we should not go forward with this PR, but I think it is mandatory that we make a plan for an effort to unify these interfaces again in the near future. By unify I also mean a single code base, but also the re-use of the same data fields for all uses, e.g. for the DoF indices: I recently had a case where the multigrid operator did not fit in cache memory although I thought it would, and the indices were a considerable issue. AMD CPUs have pretty generous caches, but it is important to not waste memory because they are not unlimited, so we should try to make data structures at least somewhat redundancy-free.

**Comment by peterrum** [comment]
> On a second thought, I believe this PR seems to replicate what is done in

Yes this is true. I am rewriting the code so that we do not need to setup a partitioner in the first step and after that the indices. This is what we discussed in person and is motivated by the fact that the determination of the indices takes the most time.

I looked into the implementation of `DoFInfo` a couple of time in the last days. I think we are soon in a state to use it (in `ConstraintInfo`). Currently, an issue is that `ConstraintInfo::read_dof_indices()` does not assumes any order of the cells while `DoFInfo::read_dof_indices()` does. But maybe we can remove this "feature" of `ConstraintInfo`. This feature is used in the setup routines of the multigrid transfer operator. However, one could presort the cells.

> Can you explain what you're trying to achieve here and what the underlying algorithm is? I see you're creating duplicated data structures (with some TODOs left in the code) to keep both unsigned int and types::global_dof_index variables in the constraint info class, in order to be able to keep some indices in global index space before a partitioner can be set up (because we cannot easily come up with a number fitting in unsigned int for arbitrary ghost indices). Is that right? 

I was storing `unsigned int` since I converted the global indices to local ones right away with the help of the partitioner. Now, we need to also store (temporally) the global indices (with an offset).

> I am wondering if there any other way to achieve the same, e.g. templating the whole class for the temporary setup where we need types::global_dof_index for the intermediate ghost data?

That would be a possibility. 

> but I think it is mandatory that we make a plan for an effort to unify these interfaces again in the near future. 

:+1:

File: include/deal.II/matrix_free/constraint_info.h [hunk 3]
```diff
@@ -81,10 +81,22 @@ namespace internal
      * user code. It combines constraint related functionalities from
      * MatrixFree and FEEvaluation.
      */
-    template <int dim, typename Number>
+    template <int dim, typename Number, typename IndexType = unsigned int>
```
Comment (peterrum) [inline include/deal.II/matrix_free/constraint_info.h:84]: @kronbichler I have introduced the class template argument as suggested.

**Review by kronbichler — APPROVED** [review]
It seems I forgot this PR while I was busy with other topics a few weeks ago. I think we can go on with this change, subject to the overall plan we discussed elsewhere.

**Pull request merged**